design/0004-multicluster-phase-2.md

```diff
@@ -0,0 +1,262 @@
+  or debug individual service mirror controllers for a specific target cluster
+  without worrying about how that will affect the mirroring for other target
+  clusters.
+* Code simplification: a single Go process no longer needs to maintain nested
+  watches where a watch on secrets starts and stops watches on target clusters.
+  Each process is responsible for a single target cluster, greatly simplifying
+  the code.
+
+The downsides of this approach are that:
+* Running more service mirror pods incurs more resource overhead, but this
+  should be negligable.
+* Running more service mirror deployments incurs more operational overhead
+  since these deployments need to be monitored and maintained.  Thankfully,
+  Prometheus makes monitoring a collection of processes just as easy as
+  monitoring a single process, assuming we use labels in a consistent way.  Any
+  observability tooling or dashboards that we build for service mirror
+  observability should have no problem dealing with metrics from multiple
+  processes.  We can also develop tooling to make upgrading easier, as discussed
+  later.
+
+## UX
+
+This proposal makes very few changes to the UX flow.
+
+
+Install the multicluster gateway, service account, and `link` CRD in both clusters.
+
+```
+linkerd --context=target multicluster install | kubectl --context=target apply -f -
+linkerd --context=source multicluster install | kubectl --context=source apply -f -
+```
+
+Link the clusters.  This will install the cluster credentails secret, `link`
+reousrce, and service mirror controller in the source cluster.
+
+```
+linkerd --context=target multicluster link --cluster-name=foobar | kubectl --context=source apply -f -
+```
+
+This will create:
+* secret/cluster-credentials-foobar
+* link.multicluster.linkerd.io/foobar
+* deployment/service-mirror-foobar
+* (also the necessary RBAC for the service mirror controller)
+
+The generated `link` resource will specify the default label selector of
+`multicluster.linkerd.io/exported=true`.
+Alternatively, users can specify their own label selector:
+
+```
+linkerd --context=target multicluster link --cluster-name=foobar -l "export-me-please=true" | kubectl --context=source apply -f -
+```
+
+Now, to actually export the service, it's just a matter of applying the correct label:
+
+```
+kubectl --context=target label svc/my-cool-svc multicluster.linkerd.io/exported=true
+```
+
+The `service-mirror-foobar` service mirror controller in the source cluster will
+see that a service in the remote cluster matches its label selector and will
+mirror that service by creating a `my-cool-svc-foobar` mirror service in the
+source cluster.
+
+Active links can be seen using `kubectl get`.  This replaces the `multicluster gateways` command:
+
+
+```
+kubectl --context=source get links
+NAME   NUM SVC   ALIVE   LATENCY_P50   LATENCY_P95   LATENCY_P99
+foobar       3    True         275ms         385ms         397ms
+```
```

Comment:
This might delve a little into the implementation design phase, but I just wanted to dig a little bit and throw some high level ideas on how this could be addressed...

As I understand it, extra printer columns are defined so that they show the values of some properties of the CRD instance, retrieved through a jsonpath expression. Stuff like the target cluster name and the services selector are readily available. But metrics like latencies would need to be actively surfaced into the CRD instance, as status properties I guess (is this what the [status subresource](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#status-subresource) is for?). The probe worker itself could update that status as it updates the latencies prometheus metrics.